src/types/proof/definition.rs

```diff
@@ -172,16 +186,188 @@ impl<H: SimpleHasher> SparseMerkleProof<H> {
+                            Some(SparseMerkleLeafNode::new(
+                                new_element_key,
+                                ValueHash::with::<H>(new_element_value),
+                            )),
+                            self.siblings,
+                        );
+
+                        // Step 3: we compare the new Merkle path against the new_root_hash
+                        new_merkle_path.unsafe_verify(new_root_hash, new_element_key)?;
+                    } else {
+                        // Case 2: The new element key is different from the leaf key (leaf creation)
+                        // Step 1: we verify the old key is going to be split following the insertion of the
+                        // new key (nonexistence proof)
+                        self.verify_nonexistence(old_root_hash, new_element_key)?;
+
+                        // Add the correct siblings of the new element key by finding the splitting nibble and
+                        // adding the default leaves to the path
+
+                        // To compute the number of default leaves we need to add, we need to:
+                        // - Compute the number of default leaves to separate the old leaf from the new leaf in the last nibble
+                        // - Compute the number of default leaves to traverse the common prefix
+                        // - Compute the number of default leaves remaining to select the former old leaf in the former last nibble
+                        // (this leaf becomes an internal node, hence the path needs to be fully specified)
+
+                        let new_key_path = NibblePath::new(new_element_key.0.to_vec());
+                        let old_key_path = NibblePath::new(leaf_node.key_hash.0.to_vec());
+
+                        // The verify_nonexistence check from before ensure that the common prefix nibbles_len is greater than the
+                        // siblings len
+                        let mut new_key_nibbles = new_key_path.nibbles();
+                        let mut old_key_nibbles = old_key_path.nibbles();
+
+                        let common_prefix_len =
+                            skip_common_prefix(&mut new_key_nibbles, &mut old_key_nibbles);
+
+                        let num_siblings = self.siblings().len();
+
+                        // We can safely unwrap these values as the check have been already performed in verify_nonexistence
+                        let mut new_key_leaf_nibble = new_key_nibbles.next().unwrap().as_usize();
+                        let mut old_key_leaf_nibble = old_key_nibbles.next().unwrap().as_usize();
+
+                        let mut default_siblings_leaf_nibble = 0;
+
+                        for h in (0..4).rev() {
+                            if !(((new_key_leaf_nibble >> h) != 0)
+                                ^ ((old_key_leaf_nibble >> h) != 0))
+                            {
```

Comment:
nit: This logic is pretty opaque. It looks like it's checking that the current bit of each key is 0? Can we simplify it by using the `bits` method of the `NibbleIterator`